### PR TITLE
refactor(useFilenamingConvention): support dynamic route filenames

### LINKED ISSUE #3465: useFilenamingConvention and file based routing catch all `[...slug]`
### Discussed in https://github.com/biomejs/biome/discussions/3444

<div type='discussions-op-text'>

<sup>Originally posted by **minht11** July 15, 2024</sup>
Many popular file based routers provide option to catch all routes.
[Next](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#catch-all-segments)
[Nuxt](https://nuxt.com/docs/guide/directory-structure/server#catch-all-route)
[Astro](https://docs.astro.build/en/guides/routing/#rest-parameters)
and I plenty more.

I might be wrong but disabling rule only for those files might not be possible because glob `"[...*].svelte"` would not work.

Given the popularity of the use case, I think we should support it by default, so users using lint rule need less config.

One option would be allowing `...` when they start inside brackets.</div>

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

Fix #3465

I had to modify the way we generate test names. To have a more robust approach, we now rely on `biome_string_case` case conversion. I had to change some test name to accommodate with this change.

## Test Plan

I added four tests.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -97,6 +97,15 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
 
   Contributed by @Conaclos
 
+- [useFilenamingConvention](https://biomejs.dev/linter/rules/use-filenaming-convention) now supports Next/Nuxt/Astro dynamic routes ([#3465](https://github.com/biomejs/biome/issues/3465)).
```
Comment (ematipico) [inline CHANGELOG.md:100]:
```suggestion
- [useFilenamingConvention](https://biomejs.dev/linter/rules/use-filenaming-convention) now supports Next.js/Nuxt/Astro dynamic routes ([#3465](https://github.com/biomejs/biome/issues/3465)).
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -97,6 +97,15 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
...
+
+  [Next](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#catch-all-segments), [Nuxt](https://nuxt.com/docs/guide/directory-structure/server#catch-all-route), and [Astro](https://docs.astro.build/en/guides/routing/#rest-parameters) support dynamic routes such as `[...slug].js` and `[[...slug]].js`.
```
Comment (ematipico) [inline CHANGELOG.md:102]:
```suggestion
  [Next.js](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes#catch-all-segments), [Nuxt](https://nuxt.com/docs/guide/directory-structure/server#catch-all-route), and [Astro](https://docs.astro.build/en/guides/routing/#rest-parameters) support dynamic routes such as `[...slug].js` and `[[...slug]].js`.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -97,6 +97,15 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
...
+
+  Biome now recognizes this syntax.
+  `slug` must contain only alphanumeric characters.
```
Comment (ematipico) [inline CHANGELOG.md:105]:
```suggestion
  Biome now recognizes this syntax. `slug` must contain only alphanumeric characters.
```

**Review by ematipico — COMMENTED** [review]
Any chance we can support SolidStart too? They use a different approach... 😓 

https://docs.solidjs.com/solid-start/building-your-application/routing#renaming-index

**Comment by ematipico** [comment]
>  All these syntaxes are so ugly to my eyes.

Ehehe, you're aren't wrong 😅

**Comment by minht11** [comment]
Just to add svelte kit also uses `+page@(app).svelte` to escape layout https://kit.svelte.dev/docs/advanced-routing#advanced-layouts-page

**Comment by Conaclos** [comment]
> +page@(

Is (app) a placeholder?

> Just to add svelte kit also uses `+page@(app).svelte` to escape layout https://kit.svelte.dev/docs/advanced-routing#advanced-layouts-page

Could you detail the syntax:
- Is `page`/`layout` fixed strings? or is it the name of the page/layout?
- If I understand corerctly, the part behind `@` is optional?
  Are `(app)`/`[id]` placeholders or are `app`/`id` placeholders?

**Comment by minht11** [comment]
Anything after `@` is just regular route name, it tells from which route above it should inherit nested layout. `+page@(app)` just means that layout should break out to `(app)` above. @ is optional and advanced feature.

I also think this should be supported only for svelte files since it is quite unusual syntax. I am not 100% garunateed, but I can't find anything for [non svelte](https://github.com/search?q=path%3A**%2F%2Bpage%40*.ts&type=code) files on search. [svelte files](https://github.com/search?q=path%3A**%2F%2Bpage%40*.svelte&type=code) using that pattern for reference.
image

**Comment by Conaclos** [comment]
I am going to merge this PR. I will consider extending the rule with a configuration to customize the accepted patterns.

**Pull request merged**